### ISSUE TITLE
sequence2sequence

### ISSUE THREAD

**Issue by imaglib** [Russell91/apollo#4]
Hi Russell,
     I found you had implemented a sequence2sequence.py to train sequence2sequence model , the structure is same as NLPCaffe, right?  I want train a model to predict target sentence given a input sentence, is it work ? Is the trained model  same to caffe model ? I did not find save and predict process in the py.

**Comment by Russell91** [comment]
Yes, the structure is the same as NLPCaffe. It's a pretty simple extension of the language model. Saving works with net.save('/path/to/file') as in the other examples. The eval_forward function still needs to be written to let the user sample sentences, but that will look pretty much like the forward() function. I encourage you to play around with the example, and if you get stuck on anything let me know and I'll see what I can do to help.

**Comment by imaglib** [comment]
Thanks!

**Comment by Russell91** [comment]
I've added a machine translation example that you may find useful as well:

https://github.com/Russell91/apollo/blob/master/examples/machine_translation.py

**Comment by imaglib** [comment]
Cool, Thanks !
I feel a little confuse about the structure for the network, both the input sentence and target sentence need to forward ?  I remember the NLP-caffe using two LSTM layers ,Can you draw the network structure?

**Comment by imaglib** [comment]
Ok , I have understand the structure. Another question,  after input sentence forward, there are two layers "hidden_seed"  and "mem_seed" , the code:
    net.forward_layer(layers.CapSequence(name='hidden_seed', sequence_lengths=lengths,
        bottoms=hidden_bottoms))
    net.forward_layer(layers.CapSequence(name='mem_seed', sequence_lengths=lengths,
        bottoms=mem_bottoms))
Why use there two layers , What is mean of  CapSequence? Why not directly connect to the target lstm layer?

**Comment by Russell91** [comment]
I've uploaded a png here: http://russellsstewart.com/s/mt/s2s.png
Note that if you pull the most recent bits the sequence2sequence example will draw this png for you by calling a new net.draw_to_file('/tmp/s2s.png') function after the forward pass.

As for your CapSequence question. I wrote that layer last week as a new trick to solve an old problem with writing sequence2sequence networks. While apollo does support variable length sequences, there is still a more fundamental problem that different sentences in the same batch may have different lengths. I've seen others deal with this by right-aligning the source sentences and left-aligning the target sentences, and then carefully pushing through zeros elsewhere to make sure you get the intended semantics. But this is all quite ugly actually, which is where the CapSequence layer comes in. It looks back at the batch of sentences and grabs the last element of each sentence, even though they might be of different lengths. This set of activations is then presented as the seed for the target-side LSTM, and everything hooks up correctly with backprop.

**Comment by imaglib** [comment]
Thanks!
